CRM/Contribute/BAO/ContributionRecur.php

```diff
@@ -586,6 +587,44 @@ public static function copyCustomValues($recurId, $targetContributionId) {
     }
   }
 
+  /**
+   * Copy custom data of the initial contribution into its recurring contributions.
+   * Returns data as an array formatted for API3 create (eg. custom_x = XX, custom_y = BB)
+   *
+   * @todo this could be made generic. also maybe it belongs in api3 somewhere. Similar to _civicrm_api3_custom_data_get()
+   *
+   * @param int $contributionID
+   *
+   * @return array
+   */
+  public static function getCustomDataForContribution($contributionID) {
+    // copy custom data
```

Comment:
Is there really no existing function? Could we not use copyGeneric?

Reply:
`copyGeneric` calls `CRM_Core_DAO::copyCustomFields()` but that's all based around objects and designed to actually copy data across. What we need in this case is a list of customfields+values in a format that api3 accepts. So that they can be handled as part of the contribution.create api3 call. I'm pretty sure there is no generic function and I'd certainly suggest the one I've created here would be better *somewhere* in api3 land - but I'm not sure where. Maybe @colemanw has ideas?
I have a similar patch for webform_civicrm as that does the custom data separately too. Generally for hooks, custom integrations etc. it's a nightmare dealing with separate entity create + customdata create because it's not easy to match them up (and it's nearly impossible to use things like civirules).

Reply:
I don't think this function does anything that this doesn't:
```
$data = civicm_api3('Contribution', 'get', ['id' => $contributionID]);
return array_filter($data, function($key) {
  return strpos($key, 'custom_') === 0;
}, ARRAY_FILTER_USE_KEY);
```

But I think we could make it even more generic than that. I'm working on a `Copy` action for APIv4.

Reply:
@colemanw Cool! I'll refactor this once the `Copy` action is ready :-)

Reply:
@mattwire what do you think of #17958